core/domain/suggestion_services.py

```diff
@@ -542,3 +556,65 @@ def check_can_resubmit_suggestion(suggestion_id, user_id):
     suggestion = get_suggestion_by_id(suggestion_id)
 
     return suggestion.author_id == user_id
+
+
+def check_can_edit_suggestion(user, suggestion_id):    # pylint: disable=too-many-return-statements
+    """Checks whether the user have permission to edit suggestion or not.
+
+    Args:
+        user: UserActionsInfo. Object having user_id, role and actions for
+            given user.
+        suggestion_id: str. Id of a suggestion.
+
+    Returns:
+        bool: Whether the user can edit the suggestion.
+    """
+
+    suggestion = get_suggestion_by_id(suggestion_id)
+    user_id = user.user_id
+    target_types = ['exploration', 'collection', 'question']
+
+    if suggestion is None:
+        raise Exception('Suggestion does not exists.')
+
+    if suggestion.is_handled:
+        raise Exception('The suggestion has already been accepted/rejected.')
+
+    if suggestion.author_id == user_id:
+        return True
+
+    if suggestion.target_type == suggestion_models.TARGET_TYPE_EXPLORATION:
+        exploration_rights = rights_manager.get_exploration_rights(
+            suggestion.target_id)
+        if exploration_rights is None:
+            return False
+
+        if rights_manager.check_can_edit_activity(
+                user, exploration_rights):
+            return True
+
+    elif suggestion.target_type == suggestion_models.TARGET_TYPE_COLLECTION:
```

Comment:
If we are doing this many checks here I don't think "Edit" I think it would be better to handle edit action separately for each activity like accept/reject. Can you please check whether that's possible?

Reply:
Yes its possible to do by creating different handler for different change command.
I have one question why did you think it will better to handel edit action separately?
In my opinion its better to make one handler because we have to just update the suggestion. We don't have to apply change command.
We were doing different checks just for checking permissions other then this there is no need for different checks of target type.

Reply:
I don't feel comfortable with the permission check you have written for editing the suggestion, Issues I see:
- Hard-coded check for each activity.
- For each activity, we have a duplicate code e.g. for edit exploration suggestion we have a duplicate code of the can_edit_exploration acl. in future we are going to have suggestions for other stuff so this structure will have more duplicate code. 
- Our exiting handle named such that it handles all actions for an activity (e.g SuggestionToExplorationActionHandler). 

I've asked you to check whether we can find a better way to do this? Please check the decorators for other suggestion handler and see how it's bulid.

Put your research somewhere and finalize something out of it, if required you can share your findings and I can help you find a better solution.